course5/homework06/homework06.cc

```diff
@@ -0,0 +1,596 @@
+    std::string operator() (const char *initStr, size_t size)
+    {
+        StringClass str(initStr, size);
+        for (int i=0; i<200; ++i)
+        {
+            str = '\0' + str;
+            str = str + '\0';
+        }
+        return std::string(str.data(), str.length());
+    }
+};
+
+template <typename StringClass>
+struct operator_plus_cstring
+{
+    static std::string name() { return "operator_plus_cstring"; }
+    std::string operator() (const char *initStr, size_t size)
+    {
+        StringClass str(initStr, size);
+        for (int i=0; i<200; ++i)
+        {
+            str = "Mokus" + str;
+            str = str + "Macska";
+        }
+        return std::string(str.data(), str.length());
+    }
+};
+
+template <typename StringClass>
+struct operator_plus_string
+{
+    static std::string name() { return "operator_plus_string"; }
+    std::string operator() (const char *initStr, size_t size)
+    {
+        StringClass str(initStr, size);
+        for (int i=0; i<200; ++i)
+        {
+            str = StringClass("Mokus") + str;
+            str = str + StringClass("Macska");
+        }
+        return std::string(str.data(), str.length());
+    }
+};
+
+template <typename StringClass>
+struct operator_plus_string_with_nul
+{
+    static std::string name() { return "operator_plus_string_with_nul"; }
+    std::string operator() (const char *initStr, size_t size)
+    {
+        StringClass str(initStr, size);
+        for (int i=0; i<200; ++i)
+        {
+            str = StringClass("Mo\0kus", 6) + str;
+            str = str + StringClass("Mac\0ska", 6);
+        }
+        return std::string(str.data(), str.length());
+    }
+};
+
+template <typename StringClass>
+struct push_back
+{
+    static std::string name() { return "push_back"; }
+    std::string operator() (const char *initStr, size_t size)
+    {
+        StringClass str(initStr, size);
+        for (int i=0; i<200; ++i)
+            str.push_back('m');
+        return std::string(str.data(), str.length());
+    }
+};
+
+template <typename StringClass>
+struct to_stream
+{
+    static std::string name() { return "to_stream"; }
+    auto operator() (const char *initStr, size_t size)
+    {
+        StringClass inputStr(initStr, size);
+        std::ostringstream os;
+        os << inputStr;
+        std::string str = os.str();
+        return std::make_tuple(
+                std::string(str.data()), // string content till closing zero at the end
+                std::string(str.data(), str.length()) // string content
+            );
+    }
+};
+
+
+auto testStrings = {
+    std::make_tuple("", 0),
+    std::make_tuple("A", 1),
+    std::make_tuple("AB", 2),
+    std::make_tuple("ABC", 3),
+    std::make_tuple("abcd", 4),
+    std::make_tuple("Mokus", 5),
+    std::make_tuple("Macska", 6),
+    std::make_tuple("Macs\0ka", 7),
+    std::make_tuple(u8"A kis mókus felmászott a fára és sok mogyorót talált, amit elrejtett, hogy legyen mit ennie a télen.", 107)
```

Comment:
It's UTF8, I counted the actual number of bytes, in this case accented characters take 2 bytes, indeed.